File: server/src/controllers/product.controller.ts
import { Product } from "@prisma/client";
import { ProductService } from "../services/product.service";
import { Request, Response } from "express";

export namespace ProductController {
  export async function getAllProducts(
    request: Request,
    response: Response<Product[]>
  ) {
    response.status(200).json(await ProductService.getAllProducts());
  }

  export async function getProductById(request: Request, response: Response) {
    response
      .status(200)
      .json(await ProductService.getProductById(parseInt(request.params.id)));
  }

  export async function createProduct(
    request: Request<Product>,
    response: Response<Product>
  ) {
    response.status(201).json(await ProductService.createProduct(request.body));
  }

  export async function updateProduct(
    request: Request,
    response: Response<Product>
  ) {
    response
      .status(200)
      .json(
        await ProductService.updateProduct(
          request.body,
          parseInt(request.params.id)
        )
      );
  }

  export async function deleteProduct(
    request: Request,
    response: Response<Product>
  ) {
    response
      .status(200)
      .json(await ProductService.deleteProduct(parseInt(request.params.id)));
  }
}

File: server/src/main/java/com/online/store/server/controllers/implementations/ProductController.java
package com.online.store.server.controllers.implementations;

import com.online.store.server.controllers.BaseController;
import com.online.store.server.payload.api.SuccessResponse;
import com.online.store.server.payload.request.ProductRequest;
import com.online.store.server.services.ProductService;
import com.online.store.server.utils.ResponseBuilder;
import jakarta.validation.Valid;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/product")
@AllArgsConstructor
public class ProductController implements BaseController<ProductRequest, Integer> {
    private ProductService productService;

    @Override
    @GetMapping
    public ResponseEntity<SuccessResponse> getAll() {
        return ResponseBuilder.createHttpSuccessResponse(
                HttpStatus.OK,
                "Success get all products.",
                productService.getAllProducts());
    }

    @Override
    @GetMapping("/{id}")
    public ResponseEntity<SuccessResponse> getById(@PathVariable Integer id) {
        return ResponseBuilder.createHttpSuccessResponse(
                HttpStatus.OK,
                "Success get product by id.",
                productService.getProductById(id));
    }

    @Override
    @PostMapping
    public ResponseEntity<SuccessResponse> create(@RequestBody @Valid ProductRequest productRequest) {
        return ResponseBuilder.createHttpSuccessResponse(
                HttpStatus.CREATED,
                "Success creating product.",
                productService.createProduct(productRequest));
    }

    @Override
    @PutMapping("/{id}")
    public ResponseEntity<SuccessResponse> update(@RequestBody @Valid ProductRequest productRequest, @PathVariable Integer id) {
        return ResponseBuilder.createHttpSuccessResponse(
                HttpStatus.OK,
                "Success updating product.",
                productService.updateProduct(productRequest, id));
    }

    @Override
    @DeleteMapping("/{id}")
    public ResponseEntity<SuccessResponse> delete(@PathVariable Integer id) {
        return ResponseBuilder.createHttpSuccessResponse(
                HttpStatus.OK,
                "Success deleting product.",
                productService.deleteProduct(id));
    }
}
